### PR TITLE
Fixes 8553: Fix contacts and ASNs missing in the search dropdown and …

### LINKED ISSUE #8553: Provider network, ASN, and contact options missins from global search selector
### NetBox version

v3.1.7

### Python version

3.8

### Steps to Reproduce

Issue 1:
1. Click on filter selections for global search
2. Look for "Contacts" in the list

Issue 2:
1. ~Search for a name in global search~
2. ~View the "assignments" column~

### Expected Behavior

Issue 1:
"Contacts" should be a selectable option, akin to to all the other searchable models

~Issue 2:~
~Number of assignments should be shown~

### Observed Behavior

Issue 1:
"Contacts" are not listed as a selectable option, although searching for a contact still works (which may or may not include various other results)

Filter List

~Issue 2:~
~Instead of the number of assignments being shown, a hyphen is displayed instead.~

### PR DESCRIPTION
**Pull request by 991jo**
…contacts not showing the number of assigned objects

### Fixes: #8553 

The contact and ASN object were not shown in the global object search dropdown.
Also the contant queryset used in the global search was not annotated with the `assigned_count`

### CODE REVIEW

File: docs/release-notes/version-3.1.md
Comment (jeremystretch) [inline docs/release-notes/version-3.1.md:8]: You can omit the changelog to avoid potential merge conflicts with other PRs.

File: netbox/netbox/constants.py
Comment (jeremystretch) [inline netbox/netbox/constants.py:189]: Change is out of scope for #8553. I've asked the OP to submit a separate bug report for this.

File: netbox/netbox/forms.py
Comment (jeremystretch) [inline netbox/netbox/forms.py:10]: While this fixes the immediate issue, it does nothing to prevent future models from also being inadvertently omitted from the list. A more complete solution would be to dynamically render the dropdown list from the `SEARCH_TYPES` dictionary. Would you like to take a shot at coming up with something?

**Review by jeremystretch — CHANGES_REQUESTED** [review]

File: netbox/netbox/constants.py
Comment (991jo) [reply netbox/netbox/constants.py:189]: okay, I will add remove that and make a new PR when the new issue is opened.

File: docs/release-notes/version-3.1.md
Comment (991jo) [reply docs/release-notes/version-3.1.md:8]: okay, I can do that. Do you add things to the changelog as an additional commit before merging or releasing or how does the "getting things into the changelog" process work?

File: netbox/netbox/constants.py
Comment (jeremystretch) [reply netbox/netbox/constants.py:189]: The second issue is #8577 FYI

File: docs/release-notes/version-3.1.md
Comment (jeremystretch) [reply docs/release-notes/version-3.1.md:8]: Whichever maintainer merges the PR is responsible for following up in the change log to ensure everything is recorded. (Admittedly it's not the best process but it works well enough with incurring any additional overhead.)

File: netbox/netbox/forms.py
Comment (991jo) [reply netbox/netbox/forms.py]: Okay, I added some code that adds a layer of hierarchy to the `constants.py` and then builds the choices from that.

**Review by jeremystretch — CHANGES_REQUESTED** [review]

File: netbox/netbox/views/__init__.py
Comment (991jo) [reply netbox/netbox/views/__init__.py:170]: I understand that the for loop is not a very nice solution.

Regarding having the hierarchy separate:
I thought the fundamental problem of this bug is that 2 partially redundant data structures have to be maintained by hand.
Separating the hierarchy from the dict containing the search types still results in 2 different places holding the information. Or did I misunderstand you?

I could keep the `SEARCH_TYPE_HIERARCHY` as is, but add code to build a flattened version for building the view. This should solve the loop problem while not having to maintain partially redundant data structures by hand.

Comment (991jo) [reply netbox/netbox/views/__init__.py]: I prepared it as described above, the `SEARCH_TYPE_HIERARCHY` still exists, and then `OBJECT_CHOICES` and `SEARCH_TYPES` are generated from that at program start.
The hierarchy is split into per-app dicts to keep in readable. That many layers of brackets and indentation would get to messy. Also the OrderedDicts are required because we might want to preserve the order in `OBJECT_CHOICES`

The result is that I don't even have to touch the code in the view and therefore not introduce a for-loop there.

Comment (991jo) [reply netbox/netbox/views/__init__.py]: @jeremystretch This PR is waiting for a response from you.

File: netbox/netbox/constants.py
Comment (jeremystretch) [inline netbox/netbox/constants.py:31]: FYI there's no need to use `OrderedDict` in this case on Python 3.7+ since dictionary ordering is now guaranteed out of the box. It's fine as-is though, I'll clean it up along with some other stuff in a project-wide sweep at some point.

**Review by jeremystretch — APPROVED** [review]

**Comment by jeremystretch** [comment]
Thanks for this!

**Pull request merged**
